modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java

```diff
@@ -232,28 +241,28 @@ public V computeIfAbsent(ICacheKey<K> key, LoadAwareCacheLoader<ICacheKey<K>, V>
         return cacheValueTuple.v1();
     }
 
-    private V compute(ICacheKey<K> key, LoadAwareCacheLoader<ICacheKey<K>, V> loader) throws Exception {
-        // Only one of the threads will succeed putting a future into map for the same key.
-        // Rest will fetch existing future and wait on that to complete.
-        CompletableFuture<Tuple<ICacheKey<K>, V>> future = completableFutureMap.putIfAbsent(key, new CompletableFuture<>());
+    private V compute(ICacheKey<K> key, LoadAwareCacheLoader<ICacheKey<K>, V> loader, CompletableFuture<Tuple<ICacheKey<K>, V>> future)
+        throws Exception {
         // Handler to handle results post processing. Takes a tuple<key, value> or exception as an input and returns
         // the value. Also before returning value, puts the value in cache.
         BiFunction<Tuple<ICacheKey<K>, V>, Throwable, Void> handler = (pair, ex) -> {
             if (pair != null) {
                 try (ReleasableLock ignore = writeLock.acquire()) {
                     onHeapCache.put(pair.v1(), pair.v2());
+                    completableFutureMap.remove(key);// Remove key from map as not needed anymore.
                 } catch (Exception e) {
                     // TODO: Catch specific exceptions to know whether this resulted from cache or underlying removal
                     // listeners/stats. Needs better exception handling at underlying layers.For now swallowing
                     // exception.
                     logger.warn("Exception occurred while putting item onto heap cache", e);
+                    completableFutureMap.remove(key);// Remove key from map as not needed anymore.
                 }
             } else {
                 if (ex != null) {
                     logger.warn("Exception occurred while trying to compute the value", ex);
                 }
+                completableFutureMap.remove(key);// Remove key from map as not needed anymore.
             }
-            completableFutureMap.remove(key); // Remove key from map as not needed anymore.
```

Comment:
We should be able to just keep `completableFutureMap.remove(key)` here instead of above 3 places. If the value is successfully added in the onHeap cache, then any other thread will read that value and future will be of no interest to that thread. If we fail to add value in heap, then the same future will be used until it is removed from the map. Until then all the threads getting the future will either get the computed value or throw exception depending on what got set in the future.

Reply:
Actually yeah. I think putting the future into map inside read lock should be enough.